core/ledger/kvledger/txmgmt/privacyenabledstate/snapshot.go

```diff
@@ -139,57 +155,261 @@ func newSnapshotWriter(
+	return &worldStateSnapshotReader{
+		pubState:       pubState,
+		pvtStateHashes: pvtStateHashes,
+	}, dbValueFormat, nil
+}
+
+func (r *worldStateSnapshotReader) Next() (*statedb.CompositeKey, []byte, error) {
+	if r.pubState != nil && r.pubState.hasMore() {
+		return r.pubState.Next()
+	}
+	if r.pvtStateHashes != nil && r.pvtStateHashes.hasMore() {
+		return r.pvtStateHashes.Next()
+	}
+	return nil, nil, nil
+}
+
+func (r *worldStateSnapshotReader) Close() {
+	if r == nil {
+		return
+	}
+	r.pubState.Close()
+	r.pvtStateHashes.Close()
+}
+
+// snapshotReader reads data from a pair of files (a data file and the corresponding metadata file)
+type snapshotReader struct {
+	dataFile *snapshot.FileReader
+	cursor   *cursor
+}
+
+func newSnapshotReader(dir, dataFileName, metadataFileName string) (*snapshotReader, byte, error) {
+	dataFilePath := filepath.Join(dir, dataFileName)
+	metadataFilePath := filepath.Join(dir, metadataFileName)
+	exist, _, err := fileutil.FileExists(dataFilePath)
+	if err != nil || !exist {
+		return nil, byte(0), err
+	}
+
+	var dataFile, metadataFile *snapshot.FileReader
+
+	defer func() {
+		if err != nil {
+			dataFile.Close()
+			metadataFile.Close()
+		}
+	}()
+
+	if dataFile, err = snapshot.OpenFile(dataFilePath, snapshotFileFormat); err != nil {
+		return nil, byte(0), errors.WithMessage(err, "error while opening data file")
+	}
+	if metadataFile, err = snapshot.OpenFile(metadataFilePath, snapshotFileFormat); err != nil {
+		return nil, byte(0), errors.WithMessage(err, "error while opening metadata file")
+	}
+	dbValueFormat, err := dataFile.DecodeBytes()
+	if err != nil {
+		return nil, byte(0), errors.WithMessage(err, "error while reading dbvalue-format")
+	}
+	if len(dbValueFormat) != 1 {
+		err = errors.Errorf("dbValueFormat is expected of length  one byte. Found [%d] length", len(dbValueFormat))
+		return nil, byte(0), err
+	}
+
+	metadata, err := loadMetadata(metadataFile)
+	if err != nil {
+		return nil, byte(0), err
+	}
+	return &snapshotReader{
+		dataFile: dataFile,
+		cursor: &cursor{
+			metadata: metadata,
+		},
+	}, dbValueFormat[0], nil
+}
+
+func loadMetadata(metadataFile *snapshot.FileReader) ([]*metadataRow, error) {
+	numMetadata, err := metadataFile.DecodeUVarInt()
+	if err != nil {
+		return nil, errors.WithMessage(err, "error while reading num-rows in metadata")
+	}
+	metadata := []*metadataRow{}
+	for i := uint64(0); i < numMetadata; i++ {
+		ns, err := metadataFile.DecodeString()
+		if err != nil {
+			return nil, errors.WithMessage(err, "error while reading namespace name")
+		}
+		numKVs, err := metadataFile.DecodeUVarInt()
+		if err != nil {
+			return nil, errors.WithMessagef(err, "error while reading num entries for the namespace [%s]", ns)
+		}
+		metadata = append(metadata, &metadataRow{
+			namespace: ns,
+			kvCounts:  numKVs,
+		})
+	}
+	return metadata, nil
+}
+
+func (r *snapshotReader) Next() (*statedb.CompositeKey, []byte, error) {
+	if !r.cursor.move() {
+		return nil, nil, nil
```

Comment:
This line becomes unreachable given that `hasMore()` is always called before calling `Next()`. Hence, it is not covered in the unit-test too. I understand that having a `hasMore()` simplifies the `Next()` method on `worldStateSnapshotReader`. At the same time, we make the `Next()` in the `snapshotReader` independent as well. Hence, I am fine with this. However, a unit-test can be added if possible.

Reply:
yes, it's unused and is there for completeness only. Will add a test to cover this.